## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

/var/folders/_5/mw87wr4s5p902pvj5xm4sqnw0000gn/T/ipykernel_97058/2050583822.py:14: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE

# contrib.head(-10)

contrib.shape

(1125659, 18)

- **1a answer here:** 

The shape matches expectations. I used contrib.head(-10) to get a list of the last 10 rows and see that that last row is 1125659 and the count of columns is 18

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE

contrib.columns

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')

- **1b answer here:** 

I can't say for sure if the 18 columsns are precisely what was in the documentation as I could not find a single bulk file with the 18 fields. However, I reviewed several files (like those listed below) and reason to believe that the data is complete as it has been blended from multiple sources and the relevant information needed to answer the question is contained.

https://www.fec.gov/data/browse-data/?tab=bulk-data

https://www.fec.gov/campaign-finance-data/candidate-committee-linkage-file-description/

https://www.fec.gov/campaign-finance-data/all-candidates-file-description/

https://www.fec.gov/campaign-finance-data/committee-master-file-description/

https://www.fec.gov/campaign-finance-data/contributions-committees-candidates-file-description/






**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [9]:
# 1c YOUR CODE HERE
contrib.head()

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here:** 

The 4 fields are cand_id (ie candidate id) which is a alphanumeric type. The cand_nm (ie candidate name) is alphabetic (unless we have a name like X Æ A-Xii, thanks to Elon). 

The contbd_st (constribution state) is a 2 digid alphabet (caps) that stands for CA (as this data is limited to california)

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE

contrib.election_tp.unique()

array(['P2016', 'G2016', nan, 'P2020'], dtype=object)

- **1d answer here:** 

I see 4 valuesP2016, G2016, P2020 and a nan). I can infer that we have two anomolies here (asusming that P2020 refers to campaign finance data for CY 2020). Also, the nan values have to be fixed as well

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [4]:
# 1e YOUR CODE HERE

#type(contrib)

contrib.info()

#contrib.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1125659 entries, 0 to 1125658
Data columns (total 18 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   cmte_id            1125659 non-null  object        
 1   cand_id            1125659 non-null  object        
 2   cand_nm            1125659 non-null  object        
 3   contbr_nm          1125659 non-null  object        
 4   contbr_city        1125633 non-null  object        
 5   contbr_st          1125659 non-null  object        
 6   contbr_zip         1125564 non-null  object        
 7   contbr_employer    967757 non-null   object        
 8   contbr_occupation  1115260 non-null  object        
 9   contb_receipt_amt  1125659 non-null  float64       
 10  contb_receipt_dt   1125659 non-null  datetime64[ns]
 11  receipt_desc       15045 non-null    object        
 12  memo_cd            144268 non-null   object        
 13  memo_text          501148 n

- **1e answer here:** 

* contbr_zip : object
* contb_receipt_amt : float64
* contb_receipt_dt: datetime64[ns]


**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [6]:
# 1f YOUR CODE HERE
contrib.info()
#contrib.isnull().sum(axis = 0)

# to get the column with the least non-null values
#contrib.count().idxmin()

# to get the column with the most non-null values
# contrib.count().idxmax()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1125659 entries, 0 to 1125658
Data columns (total 18 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   cmte_id            1125659 non-null  object        
 1   cand_id            1125659 non-null  object        
 2   cand_nm            1125659 non-null  object        
 3   contbr_nm          1125659 non-null  object        
 4   contbr_city        1125633 non-null  object        
 5   contbr_st          1125659 non-null  object        
 6   contbr_zip         1125564 non-null  object        
 7   contbr_employer    967757 non-null   object        
 8   contbr_occupation  1115260 non-null  object        
 9   contb_receipt_amt  1125659 non-null  float64       
 10  contb_receipt_dt   1125659 non-null  datetime64[ns]
 11  receipt_desc       15045 non-null    object        
 12  memo_cd            144268 non-null   object        
 13  memo_text          501148 n

- **1f answer here:** 

The max count (for non-null values) is 1125659 and there are a number of columns have the most non-null values. In order they are
* 0   cmte_id            1125659 non-null  object        
* 1   cand_id            1125659 non-null  object        
* 2   cand_nm            1125659 non-null  object        
* 3   contbr_nm          1125659 non-null  object
* 5   contbr_st          1125659 non-null  object 
* 9   contb_receipt_amt  1125659 non-null  float64       
* 10  contb_receipt_dt   1125659 non-null  datetime64[ns]
* 14  form_tp            1125659 non-null  object        
* 15  file_num           1125659 non-null  int64         
* 16  tran_id            1125659 non-null  object        
* 17  election_tp        1124234 non-null  object 

receipt_desc has the least number of non-null values (15045), implying there are 1110614, or ~ 99% of values are null. This is a candidate to be removed

I would also consider the following two columns for removal as well as they don't seem relevant to the basic question and have a ton of null values

Percentage of non-null values: 

* memo_cd: 87%
* memo_text: 55%

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [19]:
# 1g YOUR CODE HERE
# contrib.loc[contrib['cand_nm'] == 'Sanders, Bernard']

# this gives me a list of all records for Bernie
df1 = contrib.loc[contrib['cand_nm'].str.contains('Sanders, Bernard')]
# df1

# this is a lsit of all bernie records in California

df2 = df1.loc[df1['contbr_st']=='CA']
#df2

# get a list of all candidate ids for bernie in california
df2['cand_id'].unique()

array(['P60007168'], dtype=object)

- **1g answer here:** 

appears that bernie has candidate id of 'P60007168' throughout the CA dataset. No problem

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [20]:
# 1h YOUR CODE HERE
contrib['contbr_st'].unique()

array(['CA'], dtype=object)

- **1h answer here:** 

it appears that all the records are from california

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [29]:
# 1i YOUR CODE HERE

# pd.concat(x for _, x in contrib.groupby("tran_id") if len(x) > 1).head(10)

df_dupes = pd.concat(x for _, x in contrib.groupby("tran_id") if len(x) > 1)

len(df_dupes.index)

#df_dupes['cand_nm'].unique()

#df_dupes.head()

6873

In [31]:
df

In [32]:
df_dupes.head(100)

# df_dupes[df_dupes.duplicated()]

cmte_id    cand_id                  cand_nm  \
42386    C00575449  P40003576               Paul, Rand   
42835    C00575449  P40003576               Paul, Rand   
43573    C00575449  P40003576               Paul, Rand   
42428    C00575449  P40003576               Paul, Rand   
42820    C00575449  P40003576               Paul, Rand   
44679    C00575449  P40003576               Paul, Rand   
42429    C00575449  P40003576               Paul, Rand   
42821    C00575449  P40003576               Paul, Rand   
44680    C00575449  P40003576               Paul, Rand   
40218    C00575449  P40003576               Paul, Rand   
41393    C00575449  P40003576               Paul, Rand   
42836    C00575449  P40003576               Paul, Rand   
44685    C00575449  P40003576               Paul, Rand   
42356    C00575449  P40003576               Paul, Rand   
42819    C00575449  P40003576               Paul, Rand   
44710    C00575449  P40003576               Paul, Rand   
168931   C00575449  P40003576               Paul, Rand   
1071392  C00575449  P40003576               Paul, Rand   
42385    C00575449  P40003576               Paul, Rand   
42834    C00575449  P40003576               Paul, Rand   
43569    C00575449  P40003576               Paul, Rand   
40217    C00575449  P40003576               Paul, Rand   
42430    C00575449  P40003576               Paul, Rand   
42822    C00575449  P40003576               Paul, Rand   
44681    C00575449  P40003576               Paul, Rand   
128572   C00575795  P00003392  Clinton, Hillary Rodham   
708567   C00575795  P00003392  Clinton, Hillary Rodham   
243108   C00575795  P00003392  Clinton, Hillary Rodham   
575116   C00575795  P00003392  Clinton, Hillary Rodham   
243101   C00575795  P00003392  Clinton, Hillary Rodham   
483570   C00575795  P00003392  Clinton, Hillary Rodham   
128828   C00575795  P00003392  Clinton, Hillary Rodham   
709398   C00575795  P00003392  Clinton, Hillary Rodham   
130372   C00575795  P00003392  Clinton, Hillary Rodham   
576263   C00575795  P00003392  Clinton, Hillary Rodham   
244094   C00575795  P00003392  Clinton, Hillary Rodham   
707983   C00575795  P00003392  Clinton, Hillary Rodham   
397512   C00575795  P00003392  Clinton, Hillary Rodham   
566936   C00575795  P00003392  Clinton, Hillary Rodham   
406324   C00575795  P00003392  Clinton, Hillary Rodham   
573252   C00575795  P00003392  Clinton, Hillary Rodham   
245191   C00575795  P00003392  Clinton, Hillary Rodham   
708354   C00575795  P00003392  Clinton, Hillary Rodham   
397493   C00575795  P00003392  Clinton, Hillary Rodham   
569853   C00575795  P00003392  Clinton, Hillary Rodham   
393764   C00575795  P00003392  Clinton, Hillary Rodham   
570871   C00575795  P00003392  Clinton, Hillary Rodham   
397494   C00575795  P00003392  Clinton, Hillary Rodham   
644302   C00575795  P00003392  Clinton, Hillary Rodham   
409535   C00575795  P00003392  Clinton, Hillary Rodham   
646415   C00575795  P00003392  Clinton, Hillary Rodham   
407302   C00575795  P00003392  Clinton, Hillary Rodham   
568988   C00575795  P00003392  Clinton, Hillary Rodham   
406871   C00575795  P00003392  Clinton, Hillary Rodham   
570273   C00575795  P00003392  Clinton, Hillary Rodham   
245138   C00575795  P00003392  Clinton, Hillary Rodham   
575950   C00575795  P00003392  Clinton, Hillary Rodham   
393752   C00575795  P00003392  Clinton, Hillary Rodham   
573223   C00575795  P00003392  Clinton, Hillary Rodham   
245137   C00575795  P00003392  Clinton, Hillary Rodham   
710180   C00575795  P00003392  Clinton, Hillary Rodham   
129152   C00575795  P00003392  Clinton, Hillary Rodham   
576560   C00575795  P00003392  Clinton, Hillary Rodham   
409681   C00575795  P00003392  Clinton, Hillary Rodham   
644315   C00575795  P00003392  Clinton, Hillary Rodham   
407297   C00575795  P00003392  Clinton, Hillary Rodham   
646388   C00575795  P00003392  Clinton, Hillary Rodham   
406870   C00575795  P00003392  Clinton, Hillary Rodham   
661136   

- **1i answer here:** 

* yes, there are several duplicate entries (6873 to be precise)
* The interesting coincidence is that they seem to occur in 3's!
* There are others where the duplicates are in 2's where the only variable seems to be where the 
* -- contbr_occupation is 'Information Requested'

* These also are cases wehre the contributor zip code is incorrect. 

* There are other variations but these are the primary ones

---

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [58]:
# 1j YOUR CODE HERE

#contrib.info()
# contrib[contrib<0]
contrib['contb_receipt_amt'] < 0

contrib[contrib['contb_receipt_amt'] < 0]

#df3 = contrib['contb_receipt_amt'] < 0
#df3

#df4 = df3[1] = False
#df3[df3 == False].count()
df3[df3 == True].count()


#df3.info()

# np.sum((df < 0).values.ravel())

#df3['contb_receipt_amt'].count()

cmte_id    cand_id                    cand_nm  \
19       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
23       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
81       C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
190      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
213      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
...            ...        ...                        ...   
1125008  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125317  C00580399  P60008521   Christie, Christopher J.   
1125427  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
1125446  C00573519  P60005915        Carson, Benjamin S.   
1125582  C00574624  P60006111  Cruz, Rafael Edward 'Ted'   

                    contbr_nm             contbr_city contbr_st  \
19           JOLLIFF, RICHARD                   CHICO        CA   
23           JOLLIFF, RICHARD                   CHICO        CA   
81           JOLLIFF, RICHARD                   CHICO        CA   
190          NOWELL, DIANA L.  RANCHO SANTA MARGARITA        CA   
213        LICHTY, ANDREW MR.               SAN DIEGO        CA   
...                       ...                     ...       ...   
1125008  NELSON, PETER C. MR.         SAN LUIS OBISPO        CA   
1125317       ASCHER, STEPHEN                PASADENA        CA   
1125427       HANSEN, WILLIAM                GLENDALE        CA   
1125446        PECK, JOHN JR.         RANCHO SANTA FE        CA   
1125582        FARRELL, PETER                LA JOLLA        CA   

            contbr_zip                     contbr_employer  \
19      959,289,507.00                       SELF EMPLOYED   
23      959,289,507.00                       SELF EMPLOYED   
81      959,289,507.00                       SELF EMPLOYED   
190     926,884,928.00  CAPISTRAND UNIFIED SCHOOL DISTRICT   
213     921,096,720.00                       SELF EMPLOYED   
...                ...                                 ...   
1125008      934018000                       SELF EMPLOYED   
1125317      911013113                             MIRAMAR   
1125427      912081507                             RETIRED   
1125446      920670829                    PECK ENTERPRISES   
1125582      920375633                              RESMED   

            contbr_occupation  contb_receipt_amt contb_receipt_dt  \
19                    RANCHER             -25.00       2016-04-29   
23                    RANCHER            -150.00       2016-04-29   
81                    RANCHER             -60.00       2016-04-14   
190      LIBRARIAN TECHNICIAN            -100.00       2016-04-11   
213               REAL ESTATE             -25.00       2016-04-30   
...                       ...                ...              ...   
1125008               DENTIST          -2,700.00       2015-07-31   
1125317             EXECUTIVE          -2,700.00       2016-01-30   
1125427               RETIRED          -5,400.00       2015-07-01   
1125446             PRESIDENT          -2,700.00       2016-01-01   
1125582              CHAIRMAN          -2,700.00       2015-08-28   

                     receipt_desc memo_cd                 memo_text form_tp  \
19       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
23       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
81       REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
190      REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
213      REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
...                           ...     ...                       ...     ...   
1125008  REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
1125317   REATTRIBUTION TO SPOUSE       X   REATTRIBUTION TO SPOUSE   SA17A   
1125427                       NaN     NaN              CHARGED BACK   SA17A   
1125446  REDESIGNATION TO GENERAL       X  REDESIGNATION TO GENERAL   SA17A   
1125582  REDESIGNATION TO GENERAL       X 

- **1j answer here:**

* total number of negative values = 1113763

* total number of positive values = 11896

* sum of counts of negative and positive values = 1125659 (which is the total number of records)

* about 1% of thse records have a +ve contb_receipt_amt value

* it appears that the -ve values are driven by the fact that this is an accounting transaction where these amounnts are moved to a different account. You can see this by the 'receipt_desc' as 'REDESIGNATION TO GENERAL' or other reasons like 'REATTRIBUTION TO SPOUSE'

---

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [60]:
# 1k YOUR CODE HERE
contrib[(contrib['election_tp']=='P2016') & ~((contrib['contb_receipt_dt'] >= '2014-01-01') & \
                                              (contrib['contb_receipt_dt'] 
                            <= '2016-06-07'))]['contb_receipt_dt'].count()

141617

- **1k answer here:**

* the date of the donation is the field 'receipt_dt'

---

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:**

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:**

**1l.3** What columns are important?

- **1l.3 answer here:** 

**1l.4** What columns can be dropped?

- **1l.4 answer here:** 

**1l.5** What are the data problems?

- **1l.5 answer here:**

**1l.6** List any assumptions so far:

- **1l.6 answer here:**

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [ ]:
# 2a YOUR CODE HERE


**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [ ]:
# 2b YOUR CODE HERE


**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [ ]:
# 2c YOUR CODE HERE


**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [ ]:
# 2d YOUR CODE HERE


**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [ ]:
# 2e YOUR CODE HERE


- **2e answer here:**

**2f.** List any assumptions that you made up to this point:

- **2f answer here:**

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [ ]:
# 3a YOUR CODE HERE


- **3a answer here:** 

**3b.** What day(s) of the month do most people donate? (10 points)

In [ ]:
# 3b YOUR CODE HERE


- **3b answer here:** 

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k